core/node/bitswap.go

```diff
@@ -21,6 +22,7 @@ const (
 	DefaultTaskWorkerCount             = 8
 	DefaultEngineTaskWorkerCount       = 8
 	DefaultMaxOutstandingBytesPerPeer  = 1 << 20
+	DefaultProviderSearchDelay         = 20 * time.Millisecond
```

Comment:
My understanding from #8807 is that there's some analysis being done to decide on the magic number here.

In particular is there any analysis on the amount of added network traffic as a result of reducing the number here and should we expose this number in the config file?

If I'm reading the graph it correctly https://github.com/ipfs/go-ipfs/issues/8807#issuecomment-1162911032 it looks like almost no peers respond in under 100ms and some large fraction (80%?) respond by 1 second.

Dropping to 0ms, 20ms, or anything that's before Bitswap responses come back means we're sending out more requests than we do today. Understanding how many extra requests we're making vs how much time we're saving seems useful. This could also be helpful in assisting users who want to tweak the values between keeping the network chatter lower and waiting longer on Bitswap responses (e.g. gateways vs IPFS Desktop kinds of settings).

Reply:
Has anyone corroborated that 0.1% number? Seems like a pretty critical measurement. The risk of being wrong is that we'd further increase the load on DHT servers, which AFAIU is so high that many folks run in client-only mode, so I'm really loathe to further increase DHT server load.

Reply:
Hmm, we are working on a paper where we'll be able to do this, but I don't think that's going to happen in the near future. The methodology will be the same as I used for the comment, i.e., manually sending stuff on Bitswap and collecting responses. So no, I can't get this from the passive measurements we're doing at the moment, sorry...

About the 20k __positive__ responses/12 million requests number: Keep in mind that
1) I had to read those off the graph, they're not accurately from the dataset,
2) I can't ascertain the popularity of the content I tested with in a larger context, i.e., I don't know if other content is spread much more widely, generating more positive responses (or the opposite), and
3) Those 20k requests were the positive ones, but we did receive a response to almost all requests we sent out. This is because we set the `send_dont_have` flag, in order to also receive negative responses. This, critically, allows us to not rely on timeouts. (I had a number about what portion of peers respect `send_dont_have`, but I don't have it at hand. I believe  it was the vast majority.)
I suppose something like that would be possible here, too (although that's another issue): set `send_dont_have` for all requests, and start searching the DHT as soon as some portion of our neighbors responded without any positive response.

In general, I agree with 20ms being too low _as a timeout_, because it's unlikely that any meaningful number of responses will have been received. It effectively behaves as 0ms, i.e., do DHT and Bitswap discovery in parallel.

Reply:
Added `ProviderSearchDelay` to `Internal.Bitswap` config but left the default at the low value of 20ms. We could restore that to 1s and allow experiments to test the effect of changing it before enshrining a new default in code.

Reply:
For now can we expose this config but leave the default alone until at least we review @guillaumemichel's results?

Reply:
Yes. Will change to 1s